### PR TITLE
*: label external the request source via sessionvar

### PR DESCRIPTION
**Pull request by nolouch**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/44769

Problem Summary:

currently, we cannot know requests from external tools, like using the lightning-cli with `tidb` backend model to import data. 

### What is changed and how it works?

```
add a sessionvar to complement the RequestSourceType.
- external tools need start a session with setting the variable, then label the source origin
```
### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: distsql/request_builder.go [hunk 1]
```diff
@@ -306,6 +306,7 @@ func (builder *RequestBuilder) SetFromSessionVars(sv *variable.SessionVars) *Req
 	}
 	builder.RequestSource.RequestSourceInternal = sv.InRestrictedSQL
 	builder.RequestSource.RequestSourceType = sv.RequestSourceType
+	builder.RequestSource.ExplicitRequestSoureType = sv.ExplicitRequestSoureType
```
Comment (HuSharp) [inline distsql/request_builder.go:309]: Noticed some spell error with `Soure` and `reqest`
```suggestion
	builder.RequestSource.ExplicitRequestSourceType = sv.ExplicitRequestSourceType
```

File: sessionctx/variable/tidb_vars.go [hunk 2]
```diff
@@ -274,6 +274,9 @@ const (
 
 	// TiDBUseAlloc indicates whether the last statement used chunk alloc
 	TiDBUseAlloc = "last_sql_use_alloc"
+
+	// TiDBExplicitRequestSourceType indicates the source of the request, which can be ["lightning", "br", "dumping", "tispark", "background"].
```
Comment (HuSharp) [inline sessionctx/variable/tidb_vars.go:278]: Should we expose `"lightning", "br", "dumping", "tispark", "background"` as const? Maybe we need to use them both in TiDB and client-go.

Comment (nolouch) [reply sessionctx/variable/tidb_vars.go:278]: my should defined in client-go.

File: store/copr/coprocessor.go [hunk 3]
```diff
@@ -88,6 +89,7 @@ func (c *CopClient) Send(ctx context.Context, req *kv.Request, variables interfa
 	}
 	ctx = context.WithValue(ctx, tikv.TxnStartKey(), req.StartTs)
 	ctx = context.WithValue(ctx, util.RequestSourceKey, req.RequestSource)
```
Comment (glorv) [inline store/copr/coprocessor.go:91]: Seems missing the logic to update req.RequestSource from the new explicit request source

Comment (nolouch) [reply store/copr/coprocessor.go:91]: The coprocessor is set here https://github.com/pingcap/tidb/pull/44770/files#diff-4df1b0c8387d228764f0114bb4bf76ffbb54f4bbb47b45756c139b339a2d8610R276-R309

File: distsql/request_builder.go [hunk 4]
```diff
@@ -358,6 +359,12 @@ func (builder *RequestBuilder) SetResourceGroupName(name string) *RequestBuilder
 	return builder
 }
 
+// SetExplicitRequestSourceType sets the request source type.
+func (builder *RequestBuilder) SetExplicitRequestSourceType(sourceType string) *RequestBuilder {
+	builder.RequestSource.RequestSourceType = sourceType
+	return builder
+}
```
Comment (HuSharp) [inline distsql/request_builder.go:366]:
```suggestion
// SetExplicitRequestSourceType sets the explicit request source type.
func (builder *RequestBuilder) SetExplicitRequestSourceType(sourceType string) *RequestBuilder {
	builder.RequestSource.ExplicitRequestSourceType = sourceType
	return builder
}
```

File: kv/option.go [hunk 5]
```diff
@@ -188,7 +190,7 @@ const (
 	// Do not classify different tools by now.
 	InternalTxnTools = "tools"
 	// InternalTxnBR is the type of BR usage.
-	InternalTxnBR = InternalTxnTools
+	InternalTxnBR = "br"
```
Comment (HuSharp) [inline kv/option.go:193]: Should we add more internal tools?

File: sessionctx/sessionstates/session_states_test.go [hunk 6]
```diff
@@ -495,6 +495,18 @@ func TestSessionCtx(t *testing.T) {
 					`└─TableFullScan_4 10000.00 cop[tikv] table:t1 keep order:false, stats:pseudo`))
 			},
 		},
+		{
+			// check reqest source
```
Comment (HuSharp) [inline sessionctx/sessionstates/session_states_test.go:499]:
```suggestion
			// check request source
```

File: sessionctx/sessionstates/session_states_test.go [hunk 7]
```diff
@@ -495,6 +495,18 @@ func TestSessionCtx(t *testing.T) {
...
+			setFunc: func(tk *testkit.TestKit) any {
+				tk.MustExec(`set @@tidb_request_source_type="lightning"`)
+				require.Equal(t, "lightning", tk.Session().GetSessionVars().ExplicitRequestSourceType)
+				return nil
+			},
+			checkFunc: func(tk *testkit.TestKit, param any) {
+				tk.MustExec(`select count(*) from test.t1`)
```
Comment (HuSharp) [inline sessionctx/sessionstates/session_states_test.go:506]: Why need this line?

Comment (nolouch) [reply sessionctx/sessionstates/session_states_test.go:506]: I want to check if the query will influence the request source.

File: kv/option.go (hunk 5 above)
Comment (nolouch) [reply kv/option.go:193]: Here only just change what is in use.

**Review by glorv — COMMENTED** [review]
LGTM. Waiting for https://github.com/tikv/client-go/pull/868 to be merged first.

**Review by HuSharp — APPROVED** [review]
LGTM!

**Review by glorv — APPROVED** [review]
LGTM

**Review by glorv — APPROVED** [review]

**Comment by nolouch** [comment]
ptal @bb7133

**Review by qw4990 — APPROVED** [review]

**Pull request merged**